### DISCUSSION TITLE
a better approach for project-internal dependencies

### DISCUSSION THREAD

**Discussion by mberndt123** [coursier/coursier#2968] [Ideas]
### Problem Statement

I'd like to discuss a limitation of coursier's current dependency mechanism. The issue occurs in projects that publish multiple artifacts which keep a stable public API but have an unstable internal API that they use among themselves. This means that it is generally safe to upgrade from one minor version to the next for users of the public api, but not for users of the internal API, and consequently different versions of the different artifacts produced by such projects must not be mixed, because they use the internal API.
Notable examples include [Jackson](https://github.com/FasterXML/jackson) and [Pekko](https://pekko.apache.org/).

It is unfortunately very easy to create such a situation. Suppose your project depends on [`jackson-module-scala 2.13.2`](https://central.sonatype.com/artifact/com.fasterxml.jackson.module/jackson-module-scala_3/2.13.2), which depends on [`jackson-databind 2.13.2`](https://central.sonatype.com/artifact/com.fasterxml.jackson.core/jackson-databind/2.13.2). So far, so good. One day you decide to add [`avro 1.11.3`](https://central.sonatype.com/artifact/org.apache.avro/avro/1.11.3) to your dependencies. It has a dependency on [`jackson-databind 2.14.2`](https://central.sonatype.com/artifact/com.fasterxml.jackson.core/jackson-databind/2.14.2). Because `2.14.2` is newer than `2.13.2`, sbt will evict the old version of `jackson-databind` and you end up with `jackson-module-scala 2.13.2` and `jackson-databind 2.14.2` on the classpath, and now your program will crash at runtime.

@lefou suggested that this is an issue with Jackson's dependency specification: `jackson-module-scala 2.13.2` shouldn't depend on `jackson-databind 2.13.2` but on the version range `jackson-databind [2.13.2, 2.14.0)` ([original discussion on Matrix](https://matrix.to/#/!IOOrCnKIDNheemfuif:matrix.am/$nyr1yOrGD3L1vlkWlvzlH9IuZqy1X9ufK8N1g6RvB-c)) According to him, this should lead to an error in version reconciliation, giving the user an opportunity to fix the problem by adding an explicit dependency on `jackson-module-scala 2.14.2`. However, things don't seem to work this way: I've created an [sbt project](https://gist.github.com/mberndt123/2ebb7ffdcf1bc9c545e103eaedae5632) to simulate this situation, and the dependencies for the `main` project can still be resolved. (to reproduce, run `jacksonDatabindOld/publishLocal; jacksonDatabindNew/publishLocal; jacksonModuleScala/publishLocal; avro/publishLocal; show main/update` in your sbt prompt).

As far as I can tell, there is currently no way to fix this other than manually specifying all your transitive Jackson (or Akka) dependencies in the same version – but that defeats the purpose of using a dependency resolver in the first place.

### Proposed solution

I think to solve this problem it's necessary to extend the dependency model. A library `x` should have the possibility to specify in its pom file that it depends on another library `y`, and that `x` and `y` must always be the same version. This information should then be used by coursier to pull in a newer version of `x` if a newer version of `y` is pulled in as a (transitive) dependency via some other path.

Thoughts?

**Comment by durban** [comment]
Somewhat related: https://github.com/sbt/sbt/pull/7480 and https://github.com/coursier/sbt-coursier/pull/490.
